File: solutions/python3/0114.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def flatten(self, root: TreeNode) -> None:
        if not root:
            return

        self.flatten(root.right)
        self.flatten(root.left)
        root.right = self.next
        root.left = None
        self.next = root

    next = None

File: solutions/cpp/0114.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    void flatten(TreeNode* root) {
        if (!root) return;

        flatten(root->right);
        flatten(root->left);
        root->right = next;
        root->left = NULL;
        next = root;
    }

private:
    TreeNode* next = NULL;
};